content: simplify description of event-stream threat (#1292)
slsa-framework/slsa: docs/spec/draft/threats-overview.md

<tr>
<td>N/A
<td>Dependency threats (i.e. A-H, recursively)

> <td><a href="https://web.archive.org/web/20210909051737/https://schneider.dev/blog/event-stream-vulnerability-explained/">event-stream</a>: Attacker added an innocuous dependency and then later updated the dependency to add malicious behavior.

Reviewer: I don't think this change properly motivates the "How SLSA could help".  Applying SLSA recursively wouldn't necessarily help if the malicious change was submitted to GitHub.

Maybe...

```suggestion
<td><a href="https://web.archive.org/web/20210909051737/https://schneider.dev/blog/event-stream-vulnerability-explained/">event-stream</a>: Attacker added an innocuous dependency and then later updated the binary version of the dependency to add malicious behavior without updating the corresponding source code.
```

Author: Oo, I like that take. IE: SLSA doesn't actually help if your trusted publisher decides to attack you, but it can make forensics easier.

I guess we're saying that high-level SLSA Provenance techniques could make it clear that the package was built from different source?
The verification step would be: does the attested provenance agree with claimed dependency graph?

Author: either way, this change seems better than what was there before, so accepted!

> <td><a href="https://web.archive.org/web/20210909051737/https://schneider.dev/blog/event-stream-vulnerability-explained/">event-stream</a>: Attacker added an innocuous dependency and then later updated the binary version of the dependency to add malicious behavior without updating the corresponding source code.

Reviewer: It's like this:

1. Attacker has control of innocuous npm library 'foo', with the source stored in GitHub.com/attacker/foo
2. The latest version source version of foo is v1.0.0, which matches exactly the version published to npm that's also labeled v1.0.0.
3. Attacker proposes that innocent library 'bar' take a dependency on 'foo'.
4. 'bar' takes the dependency on foo. It solves some problem they have and seems fine.
5. Attacker updates GitHub.com/attacker/foo to add some basic, innocuous, fix, updates the GitHub tag to v1.0.1.
6. Attacker creates a malicious 'binary' version of foo and uploads it directly to npm as v1.0.1.
7. Dependabot (or the attacker) proposes that 'bar' update its dependency to v1.0.1
8. When bar is built it takes the malicious binary version of foo from npm.

So, I guess, if I'm reading you correctly, it means it uploads the malicious version to the package hosting location.

Any suggestions on a better way to express that?

> <td><a href="https://web.archive.org/web/20210909051737/https://schneider.dev/blog/event-stream-vulnerability-explained/">event-stream</a>: Attacker added an innocuous dependency and then later updated the binary version of the dependency to add malicious behavior without updating the corresponding source code.

Reviewer 2: So this is an xz-style attack in a way? That flow makes sense. It sounded like the content was text was implying that v1.0.0 was re-uploaded with a malicious binary version. 

I think it will be challenging to clarify this without adding too many words. Is it critical to say 
> Attacker added an innocuous dependency and then later updated

Maybe some wording like this?
```suggestion
<td><a href="https://web.archive.org/web/20210909051737/https://schneider.dev/blog/event-stream-vulnerability-explained/">event-stream</a>: Attacker controls an innocuous dependency and publishes a malicious binary version without a corresponding update to the source code.
```

It doesn't matter how the attacker controls the dependency nor does it necessarily matter who depends on this artifact. The key part of this is the ability for an attacker to insert itself into a dependency's supply chain to break the direct link from source to binary.

> <td><a href="https://web.archive.org/web/20210909051737/https://schneider.dev/blog/event-stream-vulnerability-explained/">event-stream</a>: Attacker added an innocuous dependency and then later updated the binary version of the dependency to add malicious behavior without updating the corresponding source code.

Reviewer: > It sounded like the content was text was implying that v1.0.0 was re-uploaded with a malicious binary version.

I think that could be too.  Probably shouldn't be excluded.  I left it out of my example because some package managers don't allow that (but some do!).

I like the suggestion!

> <td><a href="https://web.archive.org/web/20210909051737/https://schneider.dev/blog/event-stream-vulnerability-explained/">event-stream</a>: Attacker added an innocuous dependency and then later updated the binary version of the dependency to add malicious behavior without updating the corresponding source code.

Reviewer 3: +1 To @arewm's suggestion. I was also having a hard time clearly seeing the connection between the original text and the actual event-stream attack.

> <td>Applying SLSA recursively to all dependencies would prevent this particular vector, because the provenance would indicate that it either wasn't built from a proper builder or that the source did not come from GitHub.

Reviewer 3: Given the updated text per [Andrew's suggestion](https://github.com/slsa-framework/slsa/pull/1292/files#r2007672262), I think we may also want to clarify this explanation.
```suggestion
<td>Applying SLSA recursively to all dependencies would prevent this particular vector, because the provenance would indicate that it either wasn't built from a proper builder or that the binary did not match the source on GitHub.
```

Reviewer: yes, important distinction!
